### PR TITLE
Enhance defer load docs

### PR DESCRIPTION
**Pull request by MarcSkovMadsen**
As the title states. Builds on top of https://github.com/holoviz/panel/issues/5105 and https://github.com/holoviz/panel/pull/5107.

You can preview the document [here](https://github.com/holoviz/panel/blob/enhancement/improve-defer-load-docs/doc/how_to/callbacks/load.md).

# Todo

- [x] enhance docs
- [x] get review and improve
- [ ] Philipp: Upload assets to cdn and update links
- [ ] Finalize when #5194 has been committed

### CODE REVIEW

**Comment by MarcSkovMadsen** [comment]
## Assets

panel-longrunning-task-example
panel-defer-all-example
panel-defer-specific-example
panel-defer-onload-example
panel-defer-dependent-tasks-example

**Comment by philippjfr** [comment]
I think there is a strong over-reliance on gifs. One or two in a how-to guide seems reasonable to me, 5 seems like way too much.

File: doc/how_to/callbacks/load.md [hunk 1]
```diff
@@ -1,41 +1,189 @@
...
-Let us for example define a minimal example inside a function which we could pass to `pn.serve` (this emulates what happens when we call `panel serve` on the command line). In this example, we will create a widget without populating its options, then we will add an `onload` callback, which will set the options once the initial page is loaded.
+When you `panel serve` your app, the app is *loaded* as follows
 
-```{pyodide}
+- the app file is read and executed
+- the apps template is populated and sent to the users client (browser)
+- a web socket connection is created between the server and the users client to handle additional communication as you interact with the app.
+
+The consequence of this is that any long running code executed before the app is loaded will
+increase the waiting time for the user. **If the waiting time is more than 2-5 seconds your users might get confused and even leave the application behind**.
```
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:17]:
```suggestion
The consequence of this is that any long running code executed before the app is loaded will increase the waiting time for the user. **If the waiting time is more than 2-5 seconds your users might get confused and even leave the application behind**.
```

**Comment by MarcSkovMadsen** [comment]
> I think there is a strong over-reliance on gifs. One or two in a how-to guide seems reasonable to me, 5 seems like way too much.

I've tried to think how I can show the user the waiting time or lack of the same without gifs. I cannot figure this out. I need help. How would you do it @philippjfr ?

**Comment by MarcSkovMadsen** [comment]
I simply cannot imagine how to visually convey the deferred load without gifs @philippjfr . Please help me. How would you do it?

File: doc/how_to/callbacks/load.md [hunk 2]
```diff
@@ -1,41 +1,180 @@
...
+
+def loading_indicator():
+    return pn.indicators.LoadingSpinner(value=True, height=25, width=25, align="center")
+
+def short_running_task():
+    return "# I'm shown on load"
+
+def table(data):
+    if data is None:
+        return pn.Row(loading_indicator(), "Loading data")
```
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:157]: Should be able to simply give the loading indicator a name now.

**Comment by philippjfr** [comment]
I think it's actually okay but I do wonder if this is once again a matter of putting too much into one how to guide. How-to guides are meant to be short and to the point rather than collections of a bunch of related but slightly different approaches, i.e. I definitely don't think generators and onload are part of the same guide.

**Review by philippjfr — COMMENTED** [review]
I do think there's at least two separate guides in here

**Comment by MarcSkovMadsen** [comment]
> I do think there's at least two separate guides in here 

It is actually a tutorial instead?

**Comment by philippjfr** [comment]
We do need to define what a tutorial looks like but I don't think this is it. Just like other how to guides this material is very task oriented, i.e. it aims to get the reader how to achieve something very specific while a tutorial is much more learning oriented and gets the user to build something larger in a number of smaller steps.

**Comment by MarcSkovMadsen** [comment]
I will start splitting this into two `How-To` guides. One explaining `defer_load` another explaining `pn.state.onload`. Both aiming to explain how to defer the load.

File: doc/how_to/callbacks/load.md [hunk 3]
```diff
@@ -1,41 +1,122 @@
...
+def some_long_running_task():
+    time.sleep(5) # Some long running task
+    layout.object = "# Wow. That took some time. Are you still here?"
+
+some_long_running_task()
+
+layout.servable()
+```
+
+![panel-longrunning-task-example](https://user-images.githubusercontent.com/42288570/245752515-1329b4c3-da45-41e7-b3b5-b1b4f09eecd4.gif)
```
Comment (MarcSkovMadsen) [inline doc/how_to/callbacks/load.md:36]: Please upload this gif to the cdn @philippjfr . Thanks.

File: doc/how_to/callbacks/load.md (hunk 2 above)
Comment (MarcSkovMadsen) [reply doc/how_to/callbacks/load.md:99]: At least when #5194 has been committed to `main`.

**Comment by MarcSkovMadsen** [comment]
I've now split this into two guides. Please let me know if there is more I should do @philippjfr . Thanks.

**Comment by philippjfr** [comment]
Looks great!

File: doc/how_to/callbacks/load.md [hunk 4]
```diff
@@ -1,41 +1,122 @@
...
-pn.Row(pn.panel(render_on_load, defer_load=True))
+pn.Column(
+    short_running_task,
+    pn.bind(table, data=state.param.data),
+    pn.bind(plot, data=state.param.data),
+).servable()
 ```
 
-## Related Resources
+![panel-onload-dependent-tasks-example](https://user-images.githubusercontent.com/42288570/245752488-b2963489-bdff-4323-b801-03a763992af9.gif)
```
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:122]:
```suggestion
![panel-onload-dependent-tasks-example](https://assets.holoviz.org/panel/gifs/onload_dependent.gif)
```

File: doc/how_to/callbacks/load.md [hunk 5]
```diff
@@ -1,41 +1,122 @@
...
 
-    return widget
+pn.state.onload(some_long_running_task)
 
-# pn.serve(app) # launches the app
+layout.servable()
 ```
 
-Alternatively, we may also use the `defer_load` argument to wait to evaluate a function until the page is loaded. In a situation where page loading takes some time, a placeholder and the global `config.loading_spinner` will be displayed.
+![panel-onload-example](https://user-images.githubusercontent.com/42288570/250338232-3777ff1e-4832-4cc9-aac0-03875b8c69f5.gif)
```
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:59]:
```suggestion
![panel-onload-example](https://assets.holoviz.org/panel/gifs/onload_callback.gif)
```

File: doc/how_to/callbacks/load.md (hunk 3 above)
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:36]:
```suggestion
![panel-longrunning-task-example](https://assets.holoviz.org/panel/gifs/long_load.gif)
```

File: doc/how_to/callbacks/defer_load.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+def some_long_running_task():
+    time.sleep(5)
+    return "# Wow. That took some time. Are you still here?"
+
+
+pn.panel(some_long_running_task).servable()
+```
+
+![panel-longrunning-task-example](https://user-images.githubusercontent.com/42288570/245752515-1329b4c3-da45-41e7-b3b5-b1b4f09eecd4.gif)
```
Comment (philippjfr) [inline doc/how_to/callbacks/defer_load.md:36]:
```suggestion
![panel-longrunning-task-example](https://assets.holoviz.org/panel/gifs/long_running_task.gif)
```

File: doc/how_to/callbacks/defer_load.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    return "# I'm deferred and shown after load"
+
+
+pn.Column(
+    short_running_task,
+    pn.panel(long_running_task, defer_load=True, min_height=50, min_width=200),
+).servable()
+```
+
+![panel-defer-specific-example](https://user-images.githubusercontent.com/42288570/245752506-9ac676e9-65b2-4d9d-a01a-ce01a12dfda4.gif)
```
Comment (philippjfr) [inline doc/how_to/callbacks/defer_load.md:86]:
```suggestion
![panel-defer-specific-example](https://assets.holoviz.org/panel/gifs/defer_specific_task.gif)
```

File: doc/how_to/callbacks/defer_load.md [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+pn.extension(defer_load=True, loading_indicator=True, template="bootstrap")
+
+def long_running_task():
+    time.sleep(3)
+    return "# I'm deferred and shown after load"
+
+pn.Column("# I'm shown on load", long_running_task).servable()
+```
+
+![panel-defer-all-example](https://user-images.githubusercontent.com/42288570/245752511-b2970c4c-7144-4b1a-af36-4c90b1873de6.gif)
```
Comment (philippjfr) [inline doc/how_to/callbacks/defer_load.md:58]:
```suggestion
![panel-defer-all-example](https://assets.holoviz.org/panel/gifs/defer_all_tasks.gif)
```

File: doc/how_to/callbacks/load.md [hunk 9]
```diff
@@ -1,41 +1,122 @@
...
+
+class AppState(param.Parameterized):
+    data = param.DataFrame()
+
+    def update(self):
+        time.sleep(2)
+        state.data = pd.DataFrame({"x": [1, 2, 3, 4], "y": [1, 3, 2, 4]})
+
+def loading_indicator():
+    return pn.indicators.LoadingSpinner(value=True, height=25, width=25, align="center")
```
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:92]:
```suggestion
def loading_indicator(label):
    return pn.indicators.LoadingSpinner(
        value=True, name=label, size=25, align="center"
    )
```

File: doc/how_to/callbacks/load.md [hunk 10]
```diff
@@ -1,41 +1,122 @@
...
+
+def short_running_task():
+    return "# I'm shown on load"
+
+def table(data):
+    if data is None:
+        return pn.Row(loading_indicator(), "Loading data")
```
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:99]:
```suggestion
        return loading_indicator("Loading data")
```

File: doc/how_to/callbacks/load.md [hunk 11]
```diff
@@ -1,41 +1,122 @@
...
+
+    return pn.pane.DataFrame(data)
+
+def plot(data):
+    if data is None:
+        yield pn.Row(loading_indicator(), "Waiting for data")
```
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:105]:
```suggestion
        yield loading_indicator("Waiting for data")
```

File: doc/how_to/callbacks/load.md [hunk 12]
```diff
@@ -1,41 +1,122 @@
...
+        return
 
-```{pyodide}
+    yield pn.Row(loading_indicator(), "Transforming data")
```
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:108]:
```suggestion
    yield loading_indicator("Transforming data")
```

File: doc/how_to/callbacks/load.md (hunk 4 above)
Comment (philippjfr) [inline doc/how_to/callbacks/load.md:122]:
```suggestion
![panel-onload-dependent-tasks-example](https://assets.holoviz.org/panel/gifs/onload_dependent.gif)
```

**Pull request merged**